File: krypton/src/main/java/me/neznamy/tab/platforms/krypton/KryptonPlatform.java
package me.neznamy.tab.platforms.krypton;

import me.neznamy.tab.shared.chat.EnumChatFormat;
import me.neznamy.tab.shared.chat.IChatBaseComponent;
import me.neznamy.tab.shared.features.types.TabFeature;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.backend.BackendPlatform;
import me.neznamy.tab.shared.features.injection.PipelineInjector;
import me.neznamy.tab.shared.placeholders.expansion.TabExpansion;
import me.neznamy.tab.shared.features.nametags.NameTag;
import me.neznamy.tab.shared.placeholders.expansion.EmptyTabExpansion;
import net.kyori.adventure.text.Component;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.kryptonmc.api.Server;
import org.kryptonmc.api.entity.player.Player;

public class KryptonPlatform implements BackendPlatform {
    
    @NotNull private final KryptonTAB plugin;
    @NotNull private final Server server;

    public KryptonPlatform(@NotNull KryptonTAB plugin) {
        this.plugin = plugin;
        server = plugin.getServer();
    }

    @Override
    public void logInfo(@NotNull IChatBaseComponent message) {
        server.getConsole().sendMessage(Component.text("[TAB] ").append(message.toAdventureComponent(TAB.getInstance().getServerVersion())));
    }

    @Override
    public void logWarn(@NotNull IChatBaseComponent message) {
        server.getConsole().sendMessage(Component.text(EnumChatFormat.RED.getFormat() + "[TAB] [WARN] ").append(message.toAdventureComponent(TAB.getInstance().getServerVersion())));
    }

    @Override
    public String getServerVersionInfo() {
        return "[Krypton] " + server.getPlatform().version();
    }

    @Override
    public void registerUnknownPlaceholder(@NotNull String identifier) {
        TAB.getInstance().getPlaceholderManager().registerServerPlaceholder(identifier, -1, () -> "");
    }

    @Override
    public void loadPlayers() {
        for (Player player : server.getPlayers()) {
            TAB.getInstance().addPlayer(new KryptonTabPlayer(player));
        }
    }

    @Override
    public void registerPlaceholders() {
        new KryptonPlaceholderRegistry(this, plugin).registerPlaceholders(TAB.getInstance().getPlaceholderManager());
    }

    @Override
    public @Nullable PipelineInjector createPipelineInjector() {
        return null;
    }

    @Override
    public @NotNull NameTag getUnlimitedNameTags() {
        return new NameTag();
    }

    @Override
    public @NotNull TabExpansion createTabExpansion() {
        return new EmptyTabExpansion();
    }

    @Override
    public @Nullable TabFeature getPerWorldPlayerList() {
        return null;
    }

    @Override
    public double getTPS() {
        return -1;
    }
}

File: krypton/src/main/kotlin/me/neznamy/tab/platforms/krypton/KryptonPlatform.kt
package me.neznamy.tab.platforms.krypton

import me.neznamy.tab.api.TabFeature
import me.neznamy.tab.api.protocol.PacketBuilder
import me.neznamy.tab.platforms.krypton.features.unlimitedtags.KryptonNameTagX
import me.neznamy.tab.shared.TAB
import me.neznamy.tab.shared.backend.BackendPlatform
import me.neznamy.tab.shared.features.injection.PipelineInjector
import me.neznamy.tab.shared.placeholders.expansion.TabExpansion
import me.neznamy.tab.shared.features.nametags.NameTag
import me.neznamy.tab.shared.features.sorting.Sorting
import me.neznamy.tab.shared.placeholders.expansion.EmptyTabExpansion
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer

class KryptonPlatform(private val plugin: Main) : BackendPlatform() {

    private val server = plugin.server
    private val packetBuilder = KryptonPacketBuilder

    override fun sendConsoleMessage(message: String, translateColors: Boolean) {
        val component = if (translateColors) LegacyComponentSerializer.legacyAmpersand().deserialize(message) else Component.text(message)
        val actualMessage = Component.text().append(Component.text("[TAB] ")).append(component).build()
        server.console.sendMessage(actualMessage)
    }

    override fun registerUnknownPlaceholder(identifier: String) {
        TAB.getInstance().placeholderManager.registerServerPlaceholder(identifier, -1) { identifier }
    }

    override fun getPluginVersion(plugin: String): String? = server.pluginManager.getPlugin(plugin.lowercase())?.description?.version

    override fun loadPlayers() {
        server.players.forEach { TAB.getInstance().addPlayer(KryptonTabPlayer(it, plugin.getProtocolVersion(it))) }
    }

    override fun registerPlaceholders() {
        KryptonPlaceholderRegistry(plugin).registerPlaceholders(TAB.getInstance().placeholderManager)
    }

    override fun getPipelineInjector(): PipelineInjector {
        return KryptonPipelineInjector()
    }

    override fun getUnlimitedNametags(sorting: Sorting): NameTag {
        return KryptonNameTagX(plugin, sorting)
    }

    override fun getTabExpansion(): TabExpansion {
        return EmptyTabExpansion()
    }

    override fun getPetFix(): TabFeature? {
        return null
    }

    override fun getPerWorldPlayerlist(): TabFeature? {
        return null
    }

    override fun getPacketBuilder(): PacketBuilder {
        return packetBuilder
    }
}
